### PR TITLE
Add NOI Stepper and Step 1

### PR DESCRIPTION
**Pull request by dhaselhan**
* Add Parcels and Owners
* Change ApplicationOwnerType to OwnerType and re-use across both Apps and NOIs
* Add NOI Document controller and service for portal

### CODE REVIEW

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.html [hunk 1]
```diff
@@ -0,0 +1,68 @@
+<div>
+  <div class="step-description">
+    <h3>Identify Parcels Under Application</h3>
```
Comment (mhuseinov) [inline portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.html:3]: I guess application here used more like submission, correct?

File: services/apps/alcs/src/portal/notice-of-intent-document/notice-of-intent-document.controller.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      [VISIBILITY_FLAG.APPLICANT],
+    );
+    return this.mapPortalDocuments(documents);
+  }
+
+  @Get('/:uuid/open')
+  async open(@Param('uuid') fileUuid: string, @Req() req) {
+    const document = await this.noticeOfIntentDocumentService.get(fileUuid);
+
+    //TODO: How do we know which documents applicant can access?
```
Comment (mhuseinov) [inline services/apps/alcs/src/portal/notice-of-intent-document/notice-of-intent-document.controller.ts:66]: I wonder if this is still valid and if we have a separate ticket that captures this

File: services/apps/alcs/src/portal/notice-of-intent-submission/notice-of-intent-owner/notice-of-intent-owner.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+  it('should call through for save', async () => {
+    mockRepo.save.mockResolvedValue(new NoticeOfIntentOwner());
+
+    await service.save(new NoticeOfIntentOwner());
+
+    expect(mockRepo.save).toHaveBeenCalledTimes(1);
+  });
+
+  it('should call update for the application with the first parcels last name', async () => {
```
Comment (mhuseinov) [inline services/apps/alcs/src/portal/notice-of-intent-submission/notice-of-intent-owner/notice-of-intent-owner.service.spec.ts:217]: 'should call update for the noi with the first parcels last name'

File: services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    );
+    await queryRunner.query(
```
Comment (mhuseinov) [inline services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts:21]: Comments on tables?

**Review by mhuseinov — COMMENTED** [review]
just a few nit picks but overall looks good! 👍

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    parcel.mapAreaHectares = formData.mapArea !== undefined ? formData.mapArea : parcel.mapAreaHectares;
+    parcel.ownershipTypeCode = formData.parcelType !== undefined ? formData.parcelType : parcel.ownershipTypeCode;
+    parcel.isFarm = formData.isFarm !== undefined ? parseStringToBoolean(formData.isFarm) : parcel.isFarm;
+    parcel.purchasedDate =
+      formData.purchaseDate !== undefined ? formData.purchaseDate?.getTime() : parcel.purchasedDate;
+    parcel.isConfirmedByApplicant = formData.isConfirmedByApplicant || false;
+    parcel.crownLandOwnerType =
+      formData.crownLandOwnerType !== undefined ? formData.crownLandOwnerType : parcel.crownLandOwnerType;
+    if (formData.owners) {
+      //parcel.owners = formData.owners;
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.ts:109]: Yes, will fully support owners in next MR

File: portal-frontend/src/app/services/application-parcel/application-parcel.service.ts [hunk 6]
```diff
@@ -37,12 +34,11 @@ export class ApplicationParcelService {
     return undefined;
   }
 
-  async create(applicationSubmissionUuid: string, parcelType?: PARCEL_TYPE, ownerUuid?: string) {
+  async create(applicationSubmissionUuid: string, ownerUuid?: string) {
     try {
       return await firstValueFrom(
         this.httpClient.post<ApplicationParcelDto>(`${this.serviceUrl}`, {
           applicationSubmissionUuid,
-          parcelType,
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/services/application-parcel/application-parcel.service.ts:45]: added back, this was a mistake

File: portal-frontend/src/app/services/notice-of-intent-document/notice-of-intent-document.service.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        file,
+        documentType,
+        source,
+        `${this.serviceUrl}/application/${fileNumber}/attachExternal`
+      );
+      this.toastService.showSuccessToast('Document uploaded');
+      return res;
+    } catch (e) {
+      console.error(e);
+      this.toastService.showErrorToast('Failed to attach document to Application, please try again');
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/services/notice-of-intent-document/notice-of-intent-document.service.ts:42]: updated

File: portal-frontend/src/app/services/notice-of-intent-owner/notice-of-intent-owner.dto.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  uuid: string;
+  noticeOfIntentSubmissionUuid: string;
+  displayName: string;
+  firstName: string | null;
+  lastName: string | null;
+  organizationName: string | null;
+  phoneNumber: string | null;
+  email: string | null;
+  type: OwnerTypeDto;
+  corporateSummary?: ApplicationDocumentDto;
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/services/notice-of-intent-owner/notice-of-intent-owner.dto.ts:27]: done

File: services/apps/alcs/src/portal/notice-of-intent-document/notice-of-intent-document.controller.ts (hunk 2 above)
Comment (dhaselhan) [reply services/apps/alcs/src/portal/notice-of-intent-document/notice-of-intent-document.controller.ts:66]: definitely still valid

File: services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts (hunk 4 above)
Comment (dhaselhan) [reply services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts:21]: Will add in next MR

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.html (hunk 1 above)
Comment (dhaselhan) [reply portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.html]: nah, just forgot to update this, update to Notice of Intent

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-details.component.ts]: Updated now!

File: portal-frontend/src/app/services/notice-of-intent-owner/notice-of-intent-owner.service.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+      return undefined;
+    }
+  }
+
+  async setPrimaryContact(updateDto: SetPrimaryContactDto) {
+    try {
+      const res = await firstValueFrom(
+        this.httpClient.post<NoticeOfIntentOwnerDto>(`${this.serviceUrl}/setPrimaryContact`, updateDto)
+      );
+      this.toastService.showSuccessToast('Application saved');
```
[the remark this answers is not in the document]
Comment (dhaselhan) [reply portal-frontend/src/app/services/notice-of-intent-owner/notice-of-intent-owner.service.ts]: Done

File: services/apps/alcs/src/portal/notice-of-intent-submission/notice-of-intent-owner/notice-of-intent-owner.service.spec.ts (hunk 3 above)
Comment (dhaselhan) [reply services/apps/alcs/src/portal/notice-of-intent-submission/notice-of-intent-owner/notice-of-intent-owner.service.spec.ts]: Fixed

File: services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts (hunk 4 above)
Comment (dhaselhan) [reply services/apps/alcs/src/providers/typeorm/migrations/1691692339215-add_noi_submission_parcels_and_owners.ts:21]: Added new migration for table comments

File: portal-frontend/src/app/shared/owner-dialogs/parcel-owners/parcel-owners.component.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+  @Input() ownerService!: ApplicationOwnerService | NoticeOfIntentOwnerService;
+  @Input() documentService!: ApplicationDocumentService | NoticeOfIntentDocumentService;
+  dataSource = new MatTableDataSource([]);
+
+  @Input()
+  public set owners(owners: ApplicationOwnerDto[] | NoticeOfIntentOwnerDto[]) {
+    const sorted = owners.sort(this.sortOwners);
+    // @ts-ignore
+    this.dataSource = new MatTableDataSource(sorted);
```
Comment (mhuseinov) [inline portal-frontend/src/app/shared/owner-dialogs/parcel-owners/parcel-owners.component.ts:32]: I wonder what was the message here that you needed to use ts-ignore

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-entry/parcel-entry.component.html [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        *ngIf="enableAddNewOwner"
+        (click)="onAddNewGovernmentContact()"
+        mat-stroked-button
+        color="primary"
+        [disabled]="_disabled || parcel.owners.length > 0"
+      >
+        Add new government contact
+      </button>
+      <div>
+        <!--        <app-parcel-owners-->
```
Comment (mhuseinov) [inline portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-entry/parcel-entry.component.html:337]: I thought you were going to enable this once owners are ready

**Review by mhuseinov — APPROVED** [review]
💯

File: portal-frontend/src/app/shared/owner-dialogs/parcel-owners/parcel-owners.component.ts (hunk 10 above)
Comment (dhaselhan) [reply portal-frontend/src/app/shared/owner-dialogs/parcel-owners/parcel-owners.component.ts:32]: Something along the lines of MatTableSource<ApplicationOwnerDto[] | NoticeOfIntentOwnerDto[]> is not assignable to CdkMatTableDataSource<ApplicationOwnerDto>, for some reason the table data source hated getting passed an array of union types.

File: portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-entry/parcel-entry.component.html (hunk 11 above)
Comment (dhaselhan) [reply portal-frontend/src/app/features/notice-of-intents/edit-submission/parcels/parcel-entry/parcel-entry.component.html:337]: Good catch!

**Pull request merged**